R/regressionlinearbayesian.R

```diff
@@ -1315,3 +1317,71 @@ for sparse regression when there are more covariates than observations (Castillo
   # ggplot can't show the interaction symbol
   gsub("\u2009\u273b\u2009", " x ", name, fixed = TRUE)
 }
+
+.basregExportResiduals <- function(basregContainer, basregModel, dataset, options, ready) {
+
+  if (!ready)
+    return()
+
+  userWantsResiduals   <- options[["addResiduals"]]   && options[["residualsColumn"]]   != "" && is.null(basregContainer[["residualsColumn"]])
+  userWantsResidualSds <- options[["addResidualSds"]] && options[["residualSdsColumn"]] != "" && is.null(basregContainer[["residualSdsColumn"]])
+
+  if (!userWantsResiduals && !userWantsResidualSds)
+    return()
+
+  if (options[["summaryType"]] == "complex") {
+
+    # find the most complex model
+    mostComplexIdx <- which.max(lengths(basregModel[["which"]]))
+
+    # we copy the bas object and pretend that the most complex model is the best model,
+    # so that estimator "HPM" does exactly what we want
+    basregModelTemp <- basregModel
+    basregModelTemp[["postprobs"]][-mostComplexIdx] <- 0
+    basregModelTemp[["postprobs"]][ mostComplexIdx] <- 1
+    predictions <- predict(basregModelTemp, se.fit = userWantsResidualSds, estimator = "HPM")
+
+  } else if (options[["summaryType"]] == "median") {
+
+    # We do this for the same reason we need .basregOverwritecoefBas some weird lazy evaluation issues in R.
+    # See also https://github.com/merliseclyde/BAS/issues/56, once that is fixed we can probably remove this
+    wlsWeights <- NULL
+    if (options$wlsWeights != "") {
+      weightsVar <- options$wlsWeights
+      wlsWeights <- dataset[[weightsVar]]
+    }
+
+    basregModelTemp <- basregModel
+    basregModelTemp$call$formula <- formula(basregModel$terms)
+    basregModelTemp$call$data    <- dataset
+    basregModelTemp$call$weights <- wlsWeights
+
+    predictions <- predict(basregModelTemp, se.fit = userWantsResidualSds, estimator = "MPM")
+
+  } else {
+
+    estimator <- switch(options[["summaryType"]], best = "HPM", median = "MPM", "BMA")
+    predictions <- predict(basregModel, se.fit = userWantsResidualSds, estimator = estimator)
+
+  }
+
+  if (userWantsResiduals) {
+
+    residuals <- c(basregModel[["Y"]] - predictions[["fit"]]) # c to drop attributes
+
+    basregContainer[["residualsColumn"]] <- createJaspColumn(columnName = options[["residualsColumn"]])
```

Comment:
I was not aware this is a thing! Cool! Much better than the previous method for appending columns...

Reply:
I didn't know there was another way of doing this? Usually, I just copy other people's code when I want to do add columns to the dataset.

Reply:
We used `.setColumnDataAs*` functions before... Joris promised me a couple of years ago that he will implement a better way to do it, I just was not aware that's it's already done